File: numass-core/src/main/java/inr/numass/data/analyzers/AbstractAnalyzer.java
package inr.numass.data.analyzers;

import hep.dataforge.meta.Meta;
import hep.dataforge.tables.ListTable;
import hep.dataforge.tables.Table;
import hep.dataforge.tables.TableFormat;
import hep.dataforge.tables.TableFormatBuilder;
import inr.numass.data.api.*;
import org.jetbrains.annotations.Nullable;

import java.util.Comparator;
import java.util.stream.Stream;

import static hep.dataforge.tables.Adapters.*;
import static inr.numass.data.api.NumassPoint.HV_KEY;

/**
 * Created by darksnake on 11.07.2017.
 */
public abstract class AbstractAnalyzer implements NumassAnalyzer {
    public static String WINDOW_KEY = "window";
    public static String TIME_KEY = "timestamp";

    public static String[] NAME_LIST = {LENGTH_KEY, COUNT_KEY, COUNT_RATE_KEY, COUNT_RATE_ERROR_KEY, WINDOW_KEY, TIME_KEY};
//    public static String[] NAME_LIST_WITH_HV = {HV_KEY, LENGTH_KEY, COUNT_KEY, COUNT_RATE_KEY, COUNT_RATE_ERROR_KEY, WINDOW_KEY, TIME_KEY};
    @Nullable
    private final SignalProcessor processor;

    public AbstractAnalyzer(@Nullable SignalProcessor processor) {
        this.processor = processor;
    }

    public AbstractAnalyzer() {
        this.processor = null;
    }

    /**
     * Return unsorted stream of events including events from frames.
     * In theory, events after processing could be unsorted due to mixture of frames and events.
     * In practice usually block have either frame or events, but not both.
     *
     * @param block
     * @return
     */
    public Stream<NumassEvent> getEvents(NumassBlock block, Meta config) {
        int loChannel = config.getInt("window.lo", 0);
        int upChannel = config.getInt("window.up", Integer.MAX_VALUE);
        Stream<NumassEvent> res = getAllEvents(block).filter(event -> {
            short channel = event.getChanel();
            return channel >= loChannel && channel < upChannel;
        });
        if (config.getBoolean("sort", false)) {
            res = res.sorted(Comparator.comparing(NumassEvent::getTimeOffset));
        }
        return res;
    }

    protected Stream<NumassEvent> getAllEvents(NumassBlock block) {
        if (block.getFrames().count() == 0) {
            return block.getEvents();
        } else if (getProcessor() == null) {
            throw new IllegalArgumentException("Signal processor needed to analyze frames");
        } else {
            return Stream.concat(block.getEvents(), block.getFrames().flatMap(getProcessor()::analyze));
        }
    }

    /**
     * Get table format for summary table
     *
     * @param config
     * @return
     */
    protected TableFormat getTableFormat(Meta config) {
        return new TableFormatBuilder()
                .addNumber(HV_KEY, X_VALUE_KEY)
                .addNumber(LENGTH_KEY)
                .addNumber(COUNT_KEY)
                .addNumber(COUNT_RATE_KEY, Y_VALUE_KEY)
                .addNumber(COUNT_RATE_ERROR_KEY, Y_ERROR_KEY)
                .addColumn(WINDOW_KEY)
                .addTime()
                .build();
    }


    @Override
    public Table analyzeSet(NumassSet set, Meta config) {
        TableFormat format = getTableFormat(config);

        return new ListTable.Builder(format)
                .rows(set.getPoints().map(point -> analyzePoint(point, config)))
                .build();
    }

    @Nullable
    public SignalProcessor getProcessor() {
        return processor;
    }
}

File: numass-core/src/main/kotlin/inr/numass/data/analyzers/AbstractAnalyzer.kt
/*
 * Copyright  2017 Alexander Nozik.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package inr.numass.data.analyzers

import hep.dataforge.meta.Meta
import hep.dataforge.tables.Adapters.*
import hep.dataforge.tables.ListTable
import hep.dataforge.tables.Table
import hep.dataforge.tables.TableFormat
import hep.dataforge.tables.TableFormatBuilder
import hep.dataforge.toList
import inr.numass.data.api.NumassBlock
import inr.numass.data.api.NumassEvent
import inr.numass.data.api.NumassPoint.Companion.HV_KEY
import inr.numass.data.api.NumassSet
import inr.numass.data.api.SignalProcessor
import java.util.stream.Stream

/**
 * Created by darksnake on 11.07.2017.
 */
abstract class AbstractAnalyzer @JvmOverloads constructor(private val processor: SignalProcessor? = null) :
    NumassAnalyzer {

    /**
     * Return unsorted stream of events including events from frames.
     * In theory, events after processing could be unsorted due to mixture of frames and events.
     * In practice usually block have either frame or events, but not both.
     *
     * @param block
     * @return
     */
    override fun getEvents(block: NumassBlock, meta: Meta): List<NumassEvent> {
        val range = meta.getRange()
        return getAllEvents(block).filter { event ->
            event.amplitude.toInt() in range
        }.toList()
    }

    protected fun Meta.getRange(): IntRange {
        val loChannel = getInt("window.lo", 0)
        val upChannel = getInt("window.up", Integer.MAX_VALUE)
        return loChannel until upChannel
    }

    protected fun getAllEvents(block: NumassBlock): Stream<NumassEvent> {
        return when {
            block.frames.count() == 0L -> block.events
            processor == null -> throw IllegalArgumentException("Signal processor needed to analyze frames")
            else -> Stream.concat(block.events, block.frames.flatMap { processor.process(block, it) })
        }
    }

    /**
     * Get table format for summary table
     *
     * @param config
     * @return
     */
    protected open fun getTableFormat(config: Meta): TableFormat {
        return TableFormatBuilder()
            .addNumber(HV_KEY, X_VALUE_KEY)
            .addNumber(NumassAnalyzer.LENGTH_KEY)
            .addNumber(NumassAnalyzer.COUNT_KEY)
            .addNumber(NumassAnalyzer.COUNT_RATE_KEY, Y_VALUE_KEY)
            .addNumber(NumassAnalyzer.COUNT_RATE_ERROR_KEY, Y_ERROR_KEY)
            .addColumn(NumassAnalyzer.WINDOW_KEY)
            .addTime()
            .build()
    }


    override fun analyzeSet(set: NumassSet, config: Meta): Table {
        val format = getTableFormat(config)

        return ListTable.Builder(format)
            .rows(set.points.map { point -> analyzeParent(point, config) })
            .build()
    }

    companion object {
        val NAME_LIST = arrayOf(
            NumassAnalyzer.LENGTH_KEY,
            NumassAnalyzer.COUNT_KEY,
            NumassAnalyzer.COUNT_RATE_KEY,
            NumassAnalyzer.COUNT_RATE_ERROR_KEY,
            NumassAnalyzer.WINDOW_KEY,
            NumassAnalyzer.TIME_KEY
        )
    }
}
